### PR TITLE
feat: add OpenTelemetry metrics support

### LINKED ISSUE #461: Create the foundational metrics module using OpenTelemetry Metrics API with support for counters, histograms, and gauges
**Description:**
Create the foundational metrics module using OpenTelemetry Metrics API with support for counters, histograms, and gauges.

**Detailed Requirements:**
1. Create `mellea/telemetry/metrics.py` module with:
   - Lazy initialization of MeterProvider (similar to tracing pattern)
   - Environment-based configuration (`MELLEA_METRICS_ENABLED`)
   - Zero overhead when disabled (no-op instruments)
   - Named meter: `mellea.metrics`
2. Define base instrument creation helpers:
   - `create_counter(name, description, unit)`
   - `create_histogram(name, description, unit, boundaries)`
   - `create_up_down_counter(name, description, unit)`
3. Add dependencies to optional `[telemetry]` extra:
   - `opentelemetry-sdk` (already present)
   - Ensure metrics API available

**Files to Create:**
- `mellea/telemetry/metrics.py`

**Files to Modify:**
- `mellea/telemetry/__init__.py` - Export metrics functions
- `pyproject.toml` - Verify metrics dependencies in telemetry extra

**Environment Variables:**
| Variable | Default | Description |
|----------|---------|-------------|
| `MELLEA_METRICS_ENABLED` | `false` | Enable metrics collection |
| `MELLEA_METRICS_CONSOLE` | `false` | Print metrics to console (debug) |

**Acceptance Criteria:**
- [ ] MeterProvider initialized on first metric creation
- [ ] Instruments are no-ops when metrics disabled
- [ ] Counter, histogram, up-down-counter creation works
- [ ] No import errors when opentelemetry not installed

### PR DESCRIPTION
**Pull request by ajbozarth**
# Misc PR

## Type of PR

- [ ] Bug Fix
- [x] New Feature
- [ ] Documentation
- [ ] Other

## Description
- [x] Link to Issue: Closes #461

Implemented OpenTelemetry metrics support for Mellea:

**Changes:**
- **Refactored telemetry module**: Moved tracing implementation from `__init__.py` to dedicated `tracing.py` module for better organization
- **Added metrics module**: Implemented `mellea/telemetry/metrics.py` with full OpenTelemetry metrics support including:
  - Counter, UpDownCounter, Histogram, and Gauge metric types
  - Automatic metric provider initialization with OTLP export
  - Context-aware metric recording with span correlation
  - Comprehensive unit tests (418 lines) covering all metric types and edge cases

**Testing:**
- Added `test/telemetry/test_metrics.py` with extensive coverage of metric creation, recording, and export
- Updated existing telemetry tests to work with refactored module structure
- All tests pass with proper isolation and cleanup

The implementation follows OpenTelemetry standards and integrates seamlessly with Mellea's existing tracing infrastructure.

### Testing
- [x] Tests added to the respective file if code was changed
- [ ] New code has 100% coverage if code as added
- [x] Ensure existing tests and github automation passes (a maintainer will kick off the github automation when the rest of the PR is populated)

### CODE REVIEW

**Comment by ajbozarth** [comment]
Opening this with what I have at EOD for peer review, I want to spend more time re-reviewing this myself before final merge as it was primarily ai generated and I've only done one manual review pass so far

**Comment by ajbozarth** [comment]
I've pushed a couple commits handling review feedback from a couple more ai assistants. 

This is now ready for review, approval, and merge.

**Review by planetf1 — COMMENTED** [review]
<deleted>

File: test/telemetry/test_metrics.py [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+
+    importlib.reload(mellea.telemetry.metrics)
+    yield
+    # Reset after test
+    monkeypatch.setenv("MELLEA_METRICS_ENABLED", "false")
+    importlib.reload(mellea.telemetry.metrics)
+
+
+@pytest.fixture
+def metric_reader():
```
Comment (planetf1) [inline test/telemetry/test_metrics.py:58]: This doesn't seem to be used? 
Was the intent to use this to check values were being recorded (there are comments about needing a custom exporter - but depends at what level a check is needed. for UT this may work?)

Comment (ajbozarth) [reply test/telemetry/test_metrics.py:58]: Though this is not used here it is used in the next PR, so no need to remove it since it eventually gets used

**Comment by ajbozarth** [comment]
I have a follow up PR for #462 almost ready once this is merged (I just need to finish updating the docs before opening it)

**Comment by ajbozarth** [comment]
I tagged a few people directly to try and get this reviewed and merged to unblock the followup PR before its ready tomorrow morning

File: mellea/telemetry/metrics.py [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+    Returns:
+        Meter instance or None if metrics are disabled
+    """
+    global _meter_provider, _meter
+
+    if not _METRICS_ENABLED:
+        return None
+
+    if _meter is None:
+        _meter_provider = _setup_meter_provider()
```
Comment (psschwei) [inline mellea/telemetry/metrics.py]: The lazy initialization of `_meter_provider` and `_meter` uses globals without any locking. If two threads call `create_counter()` concurrently before `_meter` is initialized, both could race through `_setup_meter_provider()`, potentially creating duplicate `MeterProviders` and calling `metrics.set_meter_provider()` twice.

The tracing module avoids this by initializing eagerly at module load (line 85-90). Consider either:
- Doing the same eager init for metrics, or
- Adding a `threading.Lock()` around the lazy init

Comment (ajbozarth) [reply mellea/telemetry/metrics.py]: nice catch, this was caused by an error in the original issue. #461 specifically called out using lazy init like tracing, expect tracing doesn't use lazy init. This is addressed in https://github.com/generative-computing/mellea/pull/553/commits/6f2e6bc9e43c085c7e801ae104622b5cfe159d6c
